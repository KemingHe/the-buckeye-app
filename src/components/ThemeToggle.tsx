// IMPORTANT: For centralized theme management, use strictly one toggle
// component in the navbar. This avoids wrapping the entire app in a
// context provider and allows SSR.

"use client";

import { useEffect, useState } from "react";

export default function ThemeToggle(): JSX.Element {
  const [isDarkMode, setIsDarkMode] = useState<boolean>(false);

  useEffect(() => {
    try {
      const storedTheme = window.localStorage.getItem("isDarkMode");
      if (storedTheme !== null) {
        setIsDarkMode(JSON.parse(storedTheme));
      }
    } catch (error) {
      console.error("Failed to parse theme from localStorage", error);
    }
  }, []);

  // Persist the theme preference in local storage.
  useEffect(() => {
    document.documentElement.setAttribute(
      "data-theme",
      isDarkMode ? "dark" : "light"
    );
    window.localStorage.setItem("isDarkMode", JSON.stringify(isDarkMode));
  }, [isDarkMode]);

  return (
    <label
      className="flex justify-center items-center cursor-pointer space-x-2"
      htmlFor="themeToggle"
    >
      <span className="label-text">Toggle Theme</span>
      <input
        aria-label="Toggle theme"
        checked={isDarkMode}
        className="toggle theme-controller"
        id="themeToggle"
        onChange={() => setIsDarkMode(!isDarkMode)}
        type="checkbox"
      />
    </label>
  );
}
