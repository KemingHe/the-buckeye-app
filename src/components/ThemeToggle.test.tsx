import {
  type RenderResult,
  render,
  screen,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";

import ThemeToggle from "./ThemeToggle";

describe("ThemeToggle component", () => {
  it("renders the toggle theme label and checkbox", () => {
    render(<ThemeToggle />);

    // Select the label by its text.
    const labelElement = screen.getByText("Toggle Theme");
    expect(labelElement).toBeVisible();

    // Select the checkbox by its role.
    const checkboxElement = screen.getByRole("checkbox", {
      name: "Toggle theme"
    });
    expect(checkboxElement).toBeVisible();
  });

  it("toggles the theme when the checkbox is clicked", async () => {
    render(<ThemeToggle />);

    // Select the checkbox by its role.
    const checkboxElement = screen.getByRole("checkbox", {
      name: "Toggle theme"
    });

    // IMPORTANT: use waitFor to handle the asynchronous state change.

    // Click the checkbox to toggle the theme to dark.
    userEvent.click(checkboxElement);
    await waitFor(() => {
      expect(document.documentElement).toHaveAttribute("data-theme", "dark");
    });

    // Click the checkbox again to toggle the theme back to light.
    userEvent.click(checkboxElement);
    await waitFor(() => {
      expect(document.documentElement).toHaveAttribute("data-theme", "light");
    });
  });

  it("matches the snapshot", () => {
    const { asFragment }: RenderResult = render(<ThemeToggle />);
    expect(asFragment()).toMatchSnapshot();
  });
});
