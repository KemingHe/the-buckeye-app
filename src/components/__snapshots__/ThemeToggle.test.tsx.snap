// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeToggle component > matches the snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex justify-center items-center cursor-pointer space-x-2"
    for="themeToggle"
  >
    <span
      class="label-text"
    >
      Toggle Theme
    </span>
    <input
      aria-label="Toggle theme"
      class="toggle theme-controller"
      id="themeToggle"
      type="checkbox"
    />
  </label>
</DocumentFragment>
`;
